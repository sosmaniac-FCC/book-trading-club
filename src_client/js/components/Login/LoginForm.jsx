import React from 'react';

import { Link } from 'react-router-dom';

const LoginForm = (props) => (
    <form className="col s10 m4 offset-s1 offset-m4" onSubmit={props.handleLogin}>
        <div className="row">
            <div className="input-field col s10 m10 offset-s1 offset-m1">
                <input type="text" name="username" id="username" className="validate" required/>
                <label htmlFor="username" data-error="Username is a required field">Username</label>
            </div>
        </div>
        <div className="row">
            <div className="input-field col s10 m10 offset-s1 offset-m1">
    			<input type="password" name="password" id="password" className="validate" required/>
    			<label htmlFor="password" data-error="Password is a required field">Password</label>
    		</div>
    	</div>
    	<div className="row center">
    	    <input className="waves-effect waves-light btn purple accent-2 add-button-padding" type="submit" value="Login" />
    	    {"     "}
    	    <button className="waves-effect waves-light btn purple accent-1 remove-button-padding" type="button"><Link className="white-text add-button-padding" id="register-link" to={`${props.matchUrl}/CreateAccount`}>Register</Link></button>
    	</div>
    </form>
);

export default LoginForm;