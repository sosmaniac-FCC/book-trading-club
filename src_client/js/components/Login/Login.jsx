import React from 'react';

import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import { Redirect } from 'react-router-dom';

import { clearReason, registerUserState, authenticate } from '../../reducers/userReducer/userActions.jsx';
import Alert from './Alert.jsx';
import LoginForm from './LoginForm.jsx';
import Preloader from '../Preloader/Preloader.jsx';

class Login extends React.Component {
    constructor(props) {
        super(props);
        
        this.handleLogin = this.handleLogin.bind(this);
        
        this.state = {
            redirectToLogin: false  
        };
    }
    
    componentWillMount() {
        this.props.clearReason();
    }
    
    async handleLogin(e) {
        e.preventDefault();
        
        const elements = e.target.getElementsByTagName("input");
        let inputs = {};
                
        for (let i = 0; i < elements.length; i++) {
            inputs[elements[i].id] = elements[i].value;
            if (i < elements.length - 1) elements[i].value = "";
        }
        
        // form validation
        if (inputs.username != "" && inputs.password != "") {
            const authenticated = await new Promise((resolve, reject) => {
                this.props.registerUserState(resolve, reject, 'login', inputs);
            });
            
            if (!authenticated) {
                this.setState({
                    redirectToLogin: true
                });
            }
        }
    }
    
    render() {
        if (this.props.loading) {
            this.skip = false;
            return <Preloader/>;
        }
        
        if (this.state.redirectToLogin) {
            this.props.location.state = "Invalid username or password.";
        }
        
        if (this.props.token) { 
            this.props.authenticate(this.props.token); // anti-pattern error... idk, resume this tomorrow!
            
            return <Redirect to={{
                pathname: "/Book/BookSpread",
                state: "Login successful; feel free to check out these books!"
            }}/>;
        }
        
        return (
            <div className="container" id="login">
                <div className="row">
                    <h3 className="col s12 center">Sign in</h3>
                    { this.props.location.state ? <Alert message={this.props.location.state}/> : null }
                    <LoginForm handleLogin={this.handleLogin} matchUrl={this.props.match.url}/>
                </div>
            </div>
        );
    }
}

function mapStateToProps({ users }) {
    return {
        loading: users.loading,
        token: users.token
    };
}

function mapDispatchToProps(dispatch) {
    return bindActionCreators({
        clearReason,
        registerUserState,
        authenticate
    }, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(Login);