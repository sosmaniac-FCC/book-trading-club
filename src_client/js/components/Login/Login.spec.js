// expect from jest_expect.js
/* global expect */

import React from 'react';
import { shallow, configure } from 'enzyme';
import enzymeAdapter from 'enzyme-adapter-react-16';
import toJSON from 'enzyme-to-json';

configure({ adapter: new enzymeAdapter() });

import Login from './Login.jsx';

test('Login screen loads as expected.', () => {
    const component = shallow(<Login />);
    const tree = toJSON(component);
    expect(tree).toMatchSnapshot();
});

test('Login screen should have no style attributes.', () => {
    const component = shallow(<Login />);
    expect(component.find('[style]')).toHaveLength(0);
});