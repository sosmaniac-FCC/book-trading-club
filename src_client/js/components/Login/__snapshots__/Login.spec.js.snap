// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login screen loads as expected. 1`] = `
<div>
  <h1>
    Login with one's account information
  </h1>
</div>
`;
